theano/tensor/basic.py

```diff
@@ -2515,9 +2517,14 @@ def grad(self, inputs, grads):
                 axis_kept.append(i)
         gx = gz.sum(axis=axis + axis_broadcasted)
         if axis_broadcasted:
-            new_order = ['x'] * x.ndim
-            for idx, axis in enumerate(axis_kept):
-                new_order[axis] = idx
+            new_order = list(x.broadcastable)
+            idx = 0
+            for i in xrange(x.ndim):
+                if not new_order[i]:
+                    new_order[i] = idx
+                    idx += 1
+                else:
+                    new_order[i] = 'x'
```

Comment:
I'm not sure why the logic was changed here, but it looks like the result is not the same and is the reason one of the tests are failing.
The new code does not look at `axis_kept` for instance.